src/compiler/checker.ts

```diff
@@ -32791,8 +32810,38 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkNonNullAssertion(node: NonNullExpression) {
-        return node.flags & NodeFlags.OptionalChain ? checkNonNullChain(node as NonNullChain) :
-            getNonNullableType(checkExpression(node.expression));
+        const exprType = checkExpression(node.expression);
+        const resultType = node.flags & NodeFlags.OptionalChain ? checkNonNullChain(node as NonNullChain) : getNonNullableType(exprType);
+        if (!isErrorType(resultType)) {
+            addLazyDiagnostic(() => {
+                if (isTypeIdenticalTo(exprType, resultType)) {
+                    // A non-null assertion on an identifier may also suppress a used-before definition, so may still be useful even if the type is unchanged by it
+                    // Identifiers on the left-hand-side of an assignment expression, ofc, can't do this, since they're for the assigned type itself.
+                    if (isIdentifier(node.expression) && !containsUndefinedType(exprType) && !(isAssignmentExpression(node.parent) && node.parent.left === node)) {
+                        const symbol = getSymbolAtLocation(node.expression);
+                        const declaration = symbol?.valueDeclaration;
+                        if (declaration) {
+                            const isParameter = getRootDeclaration(declaration).kind === SyntaxKind.Parameter;
+                            if (!isParameter) {
+                                const flowContainer = getControlFlowContainerForIdentifier(node.expression, declaration, symbol, exprType);
+                                // We need a fake reference that isn't parented to the nonnull expression, since the non-null will affect control flow's result
+                                // by suppressing the initial return type at the end of the flow

```

Comment:
But not unprecedented! We do this for speculative/constructed control flow queries elsewhere already. See usages of `getSyntheticElementAccess` within the checker. I just couldn't reuse it because these aren't element accesses, they're just straight identifiers (sans assertion).

Reply:
> Shouldn't that be a separate concern from CFA?

🤷It's not, though. Last couple lines of `getFlowTypeOfReference` is basically "if we still have the initial type, under which contextual circumstances should we return the declared type instead", and those circumstances are basically what we wanna change here.

Reply:
@DanielRosenwasser This comment refers to a specific workaround required in the context of the control flow analysis (CFA) process. The aim here is to simulate a reference that doesn't originate from the nonnull expression to ensure that the control flow's result isn't affected solely by the presence of the non-null assertion. This workaround is reminiscent of similar approaches used in speculative or constructed control flow queries elsewhere in the codebase (like getSyntheticElementAccess within the checker). In this scenario, we're dealing with identifiers lacking assertions, and therefore, the approach isn't directly reusable. The concern addressed here is indeed entwined with the logic in the last few lines of getFlowTypeOfReference. Essentially, it delves into deciding, within certain contextual circumstances, when to return the declared type instead of retaining the initial type under which the analysis began. Hope this clarifies the context.